File: internal/permissions/permissions_test.go
package permissions

import (
	"errors"
	"fmt"
	"testing"

	"github.com/3xDevOps/Aether/internal/domain"
)

// want computes the expected Check outcome straight from the design-doc
// role table, independently of the implementation's structure.
func want(cap Capability, role domain.Role, owner, protected bool, steerOthers string) bool {
	if role == domain.RoleAdmin {
		return true
	}
	switch cap {
	case View:
		return true
	case SessionAdmin:
		return false
	case Launch, Push:
		return role == domain.RoleCollaborator
	case Handoff, Protect:
		return owner
	case Steer, Kill:
		if role != domain.RoleCollaborator {
			return false
		}
		if owner {
			return true
		}
		return !protected && steerOthers != domain.SteerOthersAdminsOnly
	}
	return false
}

// TestCheckFullMatrix exhaustively covers role x capability x ownership x
// protected x steer-others.
func TestCheckFullMatrix(t *testing.T) {
	roles := []domain.Role{domain.RoleViewer, domain.RoleCollaborator, domain.RoleAdmin}
	caps := []Capability{View, Steer, Kill, Launch, Push, Handoff, Protect, SessionAdmin}
	settings := []string{"", domain.SteerOthersAdminsOnly}
	actor := domain.MemberID("m_actor")
	other := domain.MemberID("m_other")

	for _, role := range roles {
		for _, cap := range caps {
			for _, owns := range []bool{true, false} {
				for _, protected := range []bool{false, true} {
					for _, setting := range settings {
						owner := other
						if owns {
							owner = actor
						}
						name := fmt.Sprintf("%s/%s/owns=%v/protected=%v/steer_others=%q",
							role, cap, owns, protected, setting)
						t.Run(name, func(t *testing.T) {
							err := Check(cap, Actor{ID: actor, Role: role}, Target{
								Owner:       owner,
								Protected:   protected,
								SteerOthers: setting,
							})
							allowed := err == nil
							if allowed != want(cap, role, owns, protected, setting) {
								t.Fatalf("Check = %v, want allowed=%v", err, !allowed)
							}
							if err != nil && !errors.Is(err, ErrDenied) {
								t.Fatalf("denial %v does not wrap ErrDenied", err)
							}
						})
					}
				}
			}
		}
	}
}

// TestCheckSpotRules pins the rules the matrix derives, as named examples
// mirroring the design doc's bullet points.
func TestCheckSpotRules(t *testing.T) {
	actor := domain.MemberID("m_actor")
	other := domain.MemberID("m_other")

	cases := []struct {
		name    string
		cap     Capability
		actor   Actor
		target  Target
		allowed bool
	}{
		{"viewer read-only view of another run", View,
			Actor{actor, domain.RoleViewer}, Target{Owner: other}, true},
		{"viewer cannot steer own run", Steer,
			Actor{actor, domain.RoleViewer}, Target{Owner: actor}, false},
		{"viewer cannot launch", Launch,
			Actor{actor, domain.RoleViewer}, Target{}, false},
		{"collaborator launches", Launch,
			Actor{actor, domain.RoleCollaborator}, Target{}, true},
		{"collaborator kills another's run by default", Kill,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: other}, true},
		{"admins_only blocks collaborator steering another's run", Steer,
			Actor{actor, domain.RoleCollaborator},
			Target{Owner: other, SteerOthers: domain.SteerOthersAdminsOnly}, false},
		{"admins_only keeps own runs steerable", Steer,
			Actor{actor, domain.RoleCollaborator},
			Target{Owner: actor, SteerOthers: domain.SteerOthersAdminsOnly}, true},
		{"protected run blocks non-owner kill even by default", Kill,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: other, Protected: true}, false},
		{"protected run stays open to its owner", Steer,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: actor, Protected: true}, true},
		{"admin bypasses protected and admins_only", Kill,
			Actor{actor, domain.RoleAdmin},
			Target{Owner: other, Protected: true, SteerOthers: domain.SteerOthersAdminsOnly}, true},
		{"handoff by non-owner collaborator denied", Handoff,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: other}, false},
		{"handoff by owner allowed", Handoff,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: actor}, true},
		{"protect by owner allowed", Protect,
			Actor{actor, domain.RoleCollaborator}, Target{Owner: actor}, true},
		{"session admin requires admin", SessionAdmin,
			Actor{actor, domain.RoleCollaborator}, Target{}, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := Check(tc.cap, tc.actor, tc.target)
			if (err == nil) != tc.allowed {
				t.Fatalf("Check = %v, want allowed=%v", err, tc.allowed)
			}
		})
	}
}

func TestUnknownCapabilityDenied(t *testing.T) {
	err := Check(Capability("fly"), Actor{"m", domain.RoleCollaborator}, Target{})
	if !errors.Is(err, ErrDenied) {
		t.Fatalf("unknown capability = %v, want ErrDenied", err)
	}
}

File: web/src/lib/permissions.test.ts
import {
  allowed,
  STEER_OTHERS_ADMINS_ONLY,
  type Actor,
  type RunPermission,
} from '@/lib/permissions'
import type { Member } from '@/lib/types'

// The oracle is the design-doc role table, written out again rather than
// derived from the implementation - the same shape as `want` in
// internal/permissions/permissions_test.go. The two matrices below are the
// tripwire for the mirror drifting away from the Go policy it copies.
function want(
  permission: RunPermission,
  role: Member['role'],
  owner: boolean,
  isProtected: boolean,
  steerOthers: string,
): boolean {
  if (role === 'admin') return true
  switch (permission) {
    case 'launch':
      return role === 'collaborator'
    case 'handoff':
    case 'protect':
      return owner
    case 'steer':
    case 'kill':
      if (role !== 'collaborator') return false
      if (owner) return true
      return !isProtected && steerOthers !== STEER_OTHERS_ADMINS_ONLY
  }
}

const roles: Member['role'][] = ['viewer', 'collaborator', 'admin']
const permissions: RunPermission[] = ['steer', 'kill', 'launch', 'handoff', 'protect']
const settings = ['', STEER_OTHERS_ADMINS_ONLY]
const actor = 'mem_actor'
const other = 'mem_other'

describe('the permission mirror answers as internal/permissions does', () => {
  for (const role of roles) {
    for (const permission of permissions) {
      for (const owns of [true, false]) {
        for (const isProtected of [false, true]) {
          for (const steerOthers of settings) {
            const name = `${role}/${permission}/owns=${owns}/protected=${isProtected}/steer_others="${steerOthers}"`
            it(name, () => {
              expect(
                allowed(
                  permission,
                  { id: actor, role },
                  {
                    owner: owns ? actor : other,
                    protected: isProtected,
                    steerOthers,
                  },
                ),
              ).toBe(want(permission, role, owns, isProtected, steerOthers))
            })
          }
        }
      }
    }
  }
})

// The named rules the matrix derives, mirroring TestCheckSpotRules.
test.each<[string, RunPermission, Actor, Parameters<typeof allowed>[2], boolean]>([
  ['viewer cannot steer own run', 'steer', { id: actor, role: 'viewer' }, { owner: actor }, false],
  ['viewer cannot launch', 'launch', { id: actor, role: 'viewer' }, {}, false],
  ['collaborator launches', 'launch', { id: actor, role: 'collaborator' }, {}, true],
  [
    'collaborator kills another run by default',
    'kill',
    { id: actor, role: 'collaborator' },
    { owner: other },
    true,
  ],
  [
    'admins_only blocks a collaborator steering another run',
    'steer',
    { id: actor, role: 'collaborator' },
    { owner: other, steerOthers: STEER_OTHERS_ADMINS_ONLY },
    false,
  ],
  [
    'admins_only keeps own runs steerable',
    'steer',
    { id: actor, role: 'collaborator' },
    { owner: actor, steerOthers: STEER_OTHERS_ADMINS_ONLY },
    true,
  ],
  [
    'a protected run blocks a non-owner kill even by default',
    'kill',
    { id: actor, role: 'collaborator' },
    { owner: other, protected: true },
    false,
  ],
  [
    'a protected run stays open to its owner',
    'steer',
    { id: actor, role: 'collaborator' },
    { owner: actor, protected: true },
    true,
  ],
  [
    'an admin bypasses protected and admins_only',
    'kill',
    { id: actor, role: 'admin' },
    { owner: other, protected: true, steerOthers: STEER_OTHERS_ADMINS_ONLY },
    true,
  ],
  [
    'handoff by a non-owner collaborator is denied',
    'handoff',
    { id: actor, role: 'collaborator' },
    { owner: other },
    false,
  ],
  [
    'protect by the owner is allowed',
    'protect',
    { id: actor, role: 'collaborator' },
    { owner: actor },
    true,
  ],
])('%s', (_name, permission, who, target, expected) => {
  expect(allowed(permission, who, target)).toBe(expected)
})

// The one question the Go policy never has to answer: the dashboard renders
// before server.info arrives, so an unknown role is optimistic and the server
// refuses anything this member may not do.
test('an unknown role is allowed everything, because the server still checks', () => {
  for (const permission of permissions) {
    expect(allowed(permission, { id: null, role: null }, { owner: other })).toBe(true)
  }
})

// An empty actor id never counts as ownership, so an unidentified caller
// cannot inherit a run whose owner field is also empty.
test('an empty actor id owns nothing', () => {
  expect(allowed('handoff', { id: '', role: 'collaborator' }, { owner: '' })).toBe(false)
})
